### PR TITLE
[SNMP]: SONiC SNMP Changes to support IPv6

### PR DESCRIPTION
**Pull request by SuvarnaMeenakshi**
Document  captures the changes required to support SNMP over IPv6 for single asic platforms.
1. [SNMP][IPv6]: Fix SNMP IPv6 reachability issue in certain scenarios https://github.com/sonic-net/sonic-buildimage/pull/15487
2. [SNMP][IPv6]: Fix to use link local IPv6 address as snmp agentAddress https://github.com/sonic-net/sonic-buildimage/pull/16013
3. [SNMP]: Modify minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG https://github.com/sonic-net/sonic-buildimage/pull/17045

### CODE REVIEW

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (venkatmahalingam) [inline doc/snmp/snmp-changes-to-support-ipv6.md:71]: Changes done to workaround the issue?

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (venkatmahalingam) [inline doc/snmp/snmp-changes-to-support-ipv6.md:73]: Hope the current behavior to listen on any IP by default will be retained as is and change to listen on particular interface if needed as a config option.

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (venkatmahalingam) [inline doc/snmp/snmp-changes-to-support-ipv6.md:55]: Any possibility of adding a patch to net-snmp to store the source IP of the request for IPv6 case also?

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md:73]: current behavior is :
1. Use SNMP_AGENT_ADDRESS_CONFIG configuration in config_db if present.
2. Else, Listen on Mgmt and loopback IP present in config_db
3. Else listen on any IP.

Previously:
1. Use SNMP_AGENT_ADDRESS_CONFIG configuration in config_db if present.
2. Else listen on any IP.

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md:55]: That is a possible solution but that will be a move involved change in snmpd IPv6 related files.

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (venkatmahalingam) [inline doc/snmp/snmp-changes-to-support-ipv6.md:117]: This new change may not work for all SONiC users, can introduce backward compatibility issues as we dont have any issue with IPv4 case.

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (venkatmahalingam) [inline doc/snmp/snmp-changes-to-support-ipv6.md:126]: Hope you are fixing this as part of the initial pull request itself, otherwise we're breaking the working feature when the DHCP is desired for the mgmt interface.

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (venkatmahalingam) [reply doc/snmp/snmp-changes-to-support-ipv6.md:73]: When you say current behavior, were the code changes all merged? I thought, we're informing about the "Listen on Mgmt and loopback IP present in config_db" to the community before making the change.

Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md:73]: We have this change merged in master branch and recent branches to have Ipv6  support and have enhanced security of snmpd not listening on all IPs.

**Comment by zhangyanzhao** [comment]
community review recording https://zoom.us/rec/share/IhwpNF2czHBDY1OeZwnKdTmnPr4gc2AmItBcRYc1HgFUJG539cPaeIsvdDMaeli3.AqXFYTMBt18iU1XQ

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (lguohan) [inline doc/snmp/snmp-changes-to-support-ipv6.md:33]: why the source IP is the issue? in snmp pdu, they have request-id, and response has to match the request id.

from rfc 1157.

   GetResponse-PDU represents the name and value of that object whose
   name is, in the lexicographical ordering of the names of all objects
   available for get operations in the relevant MIB view, together with
   the value of the name field of the given component, the immediate
   successor to that value.  The value of the error-status field of the
   GetResponse-PDU is noError and the value of the errorindex field is
   zero.  The value of the request-id field of the GetResponse-PDU is
   that of the received message.

**Review by venkatmahalingam — CHANGES_REQUESTED** [review]
We should make sure that the backward compatibility issues are addressed properly before merging this HLD.

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (venkatmahalingam) [reply doc/snmp/snmp-changes-to-support-ipv6.md:55]: Can we look at adding the patch in the net-snmp?

**Comment by SuvarnaMeenakshi** [comment]
@venkatmahalingam , @dgsudharsan , thank you for reviewing.
Based on the feedback, one approach that I would like your input for is instead of modifying snmpd.conf.j2 default behavior to get the MGMT and Loopback IP, if we keep the snmpd.conf.j2 as it was and modify the minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG config table with MGMT and LOOPBACK IP in config_db.
This way, if we load a minigraph with Management and Looopback IP, that will get used and snmpd will listen on those IPs.
If we use config_db, then snmpd will listen on IPs in SNMP_AGENT_ADDRESS_CONFIG  if that table is present, if not it will listen on any IP. This was the initial behavior. 
Here is a diff for the proposal https://github.com/SuvarnaMeenakshi/sonic-buildimage/commit/d1a9fa38710116b1b0d5e43a7c29bb93c85a19f4

**Comment by zhangyanzhao** [comment]
@SuvarnaMeenakshi  can you please add the code PRs to this HLD? Thanks.

**Comment by dgsudharsan** [comment]
> @venkatmahalingam , @dgsudharsan , @qiluo-msft , thank you for reviewing. Based on the feedback, one approach that I would like your input for is instead of modifying snmpd.conf.j2 default behavior to get the MGMT and Loopback IP, if we keep the snmpd.conf.j2 as it was and modify the minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG config table with MGMT and LOOPBACK IP in config_db. This way, if we load a minigraph with Management and Looopback IP, that will get used and snmpd will listen on those IPs. If we use config_db, then snmpd will listen on IPs in SNMP_AGENT_ADDRESS_CONFIG if that table is present, if not it will listen on any IP. This was the initial behavior. Here is a diff for the proposal [SuvarnaMeenakshi/sonic-buildimage@d1a9fa3](https://github.com/SuvarnaMeenakshi/sonic-buildimage/commit/d1a9fa38710116b1b0d5e43a7c29bb93c85a19f4)

Hi @SuvarnaMeenakshi 
If I understand, this flow will get executed only when someone calls config load minigraph. If there is an upgrade flow, this will not get executed, so backward compatibility is maintained. Similarly if someone creates config_db.json directly and performs reboot, this flow will not get triggered. Can you confirm this?

**Comment by SuvarnaMeenakshi** [comment]
> rly if someone creates config_db.json directly and perf



> > @venkatmahalingam , @dgsudharsan , @qiluo-msft , thank you for reviewing. Based on the feedback, one approach that I would like your input for is instead of modifying snmpd.conf.j2 default behavior to get the MGMT and Loopback IP, if we keep the snmpd.conf.j2 as it was and modify the minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG config table with MGMT and LOOPBACK IP in config_db. This way, if we load a minigraph with Management and Looopback IP, that will get used and snmpd will listen on those IPs. If we use config_db, then snmpd will listen on IPs in SNMP_AGENT_ADDRESS_CONFIG if that table is present, if not it will listen on any IP. This was the initial behavior. Here is a diff for the proposal [SuvarnaMeenakshi/sonic-buildimage@d1a9fa3](https://github.com/SuvarnaMeenakshi/sonic-buildimage/commit/d1a9fa38710116b1b0d5e43a7c29bb93c85a19f4)
> 
> Hi @SuvarnaMeenakshi If I understand, this flow will get executed only when someone calls config load minigraph. If there is an upgrade flow, this will not get executed, so backward compatibility is maintained. Similarly if someone creates config_db.json directly and performs reboot, this flow will not get triggered. Can you confirm this?

During upgrade_flow, if the config_db.json file is present, then backward compatibility will be maintained.
Yes, if someone is using config_db.json directly and performing a reboot then this will not be triggered and snmp will listen on any IP.
Only when load_minigraph is issued this flow will be triggered and snmp will listen on loopback and mgmt IPs configured in the minigraph.xml.

**Comment by venkatmahalingam** [comment]
> modify the minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG config table with MGMT and LOOPBACK IP in config_db.

Thanks @SuvarnaMeenakshi, this looks good.

**Comment by SuvarnaMeenakshi** [comment]
> > modify the minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG config table with MGMT and LOOPBACK IP in config_db.
> 
> Thanks @SuvarnaMeenakshi, this looks good.

@dgsudharsan @venkatmahalingam thank you, I have code PR here https://github.com/sonic-net/sonic-buildimage/pull/17045

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md]: Modified as suggested

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md]: Modified to keep the behavior as is, only when using miningraph.xml to load initial configuration, the SNMP_AGENT_ADDRESS_CONFIG will be updated and snmpd will listen on mgmt/loopback0 IPs

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md]: modified to be backward compatible

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md]: Not required with current suggested change

**Comment by SuvarnaMeenakshi** [comment]
> @SuvarnaMeenakshi can you please add the code PRs to this HLD? Thanks.

Updated PR in the document.

> > > modify the minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG config table with MGMT and LOOPBACK IP in config_db.
> > 
> > 
> > Thanks @SuvarnaMeenakshi, this looks good.
> 
> @dgsudharsan @venkatmahalingam thank you, I have code PR here [sonic-net/sonic-buildimage#17045](https://github.com/sonic-net/sonic-buildimage/pull/17045)



> > > modify the minigraph parser to update SNMP_AGENT_ADDRESS_CONFIG config table with MGMT and LOOPBACK IP in config_db.
> > 
> > 
> > Thanks @SuvarnaMeenakshi, this looks good.
> 
> @dgsudharsan @venkatmahalingam thank you, I have code PR here [sonic-net/sonic-buildimage#17045](https://github.com/sonic-net/sonic-buildimage/pull/17045)

@dgsudharsan @venkatmahalingam Please help review this document PR

**Review by venkatmahalingam — APPROVED** [review]

**Comment by SuvarnaMeenakshi** [comment]
@qiluo-msft please help review and merge this.

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (qiluo-msft) [inline doc/snmp/snmp-changes-to-support-ipv6.md:8]: Could you rephase "SNMP over **IPv6** should be successful over both **IPv4** and IPv6"?

File: doc/snmp/snmp-changes-to-support-ipv6.md
Comment (qiluo-msft) [inline doc/snmp/snmp-changes-to-support-ipv6.md:138]: Since this issue is only on single asic, why fixing multi asic in this PR?

Comment (SuvarnaMeenakshi) [reply doc/snmp/snmp-changes-to-support-ipv6.md:139]: Yes, only on single-asic, not fixing for multi-asic in this PR.

**Comment by SuvarnaMeenakshi** [comment]
@qiluo-msft  can you help merge this PR?
We have approval from Dell team.

**Comment by SuvarnaMeenakshi** [comment]
@dgsudharsan can you approve the doc PR

**Comment by SuvarnaMeenakshi** [comment]
> @dgsudharsan can you approve the doc PR

Merging the PR based on discussion in the PR, thanks.

**Comment by zhangyanzhao** [comment]
back ported to previous releases as well

**Comment by skg-net** [comment]
@SuvarnaMeenakshi Can you please update the Quality Metric (Alpha/Beta/GA) for the feature either in this PR comments or in HLD itself based on https://github.com/sonic-net/SONiC/blob/master/doc/SONiC%20feature%20quality%20definition.md
Thanks

**Comment by SuvarnaMeenakshi** [comment]
> @SuvarnaMeenakshi Can you please update the Quality Metric (Alpha/Beta/GA) for the feature either in this PR comments or in HLD itself based on https://github.com/sonic-net/SONiC/blob/master/doc/SONiC%20feature%20quality%20definition.md Thanks

Sure, will check the quality metric and add the required details. Is this a new additional requirement for new features or changes?

**Pull request merged**
